skimage/morphology/rolling_ball.py

```diff
@@ -0,0 +1,197 @@
+import numpy as np
+from numpy.lib.stride_tricks import as_strided
+
+from skimage.util import invert, view_as_windows
+from ._rolling_ball_cy import apply_kernel, apply_kernel_nan
+
+
+def rolling_ellipsoid(image, kernel_shape=100, intensity_vertex=100,
+                      has_nan=False, num_threads=None):
+    """
+    Estimate background intensity using a rolling ellipsoid.
+
+    The rolling ellipsoid algorithm estimates background intensity for a
+    grayscale image in case of uneven exposure. It is a generalization of the
+    frequently used rolling ball algorithm [1]_.
+
+    Parameters
+    ----------
+    image : ndarray
+        The image to be filtered.
+    kernel_shape: ndarray or scalar, optional
+        The length of the non-intensity vertices of the ellipsoid. If
+        ``kernel_shape`` is a ndarray, it must have the same dimensions as
+        ``image``. If ``kernel_shape`` is a scalar it will be extended to
+        have the same dimension via
+        ``kernel_shape = kernel_shape * np.ones_like(image.shape)``.
+        All elements must be greater than 0.
+    intensity_vertex : scalar, optional
+        The length of the intensity vertex of the ellipsoid. Must be greater
+        than 0.
+    has_nan: bool, optional
```

Comment:
We've had a few discussions in the repo about how best to handle nans. None of them included `has_nan=` 😂 Suggestions include `nansafe=False`, and also *not* providing a nansafe option at all, but rather informing the user of how to proceed with an image that contains nans. (I this case, I think replacing them by inf and then bringing them back would work?)

I am happy to go with `nansafe=` for now, since the implementation is done. But, perhaps the least controversial choice is to *not* include this in the API, but keep the Cython implementation, and defer implementation of nansafe until we have decided on how to consistently handle nans throughout the library. It is very easy to add keyword arguments, but much harder to take them back!

Reply:
It depends on what you want to happen if the image contains `NaN`. You seem to suggest eating them, in which case you could replace them with `np.Inf`, `0` or `np.iinfo(img.dtype).maxval` depending on the image's datatype and if it is a white/black background.

You can't simply replace them and bring them back to propagate. You'd have to set all affected pixels in the `kernel.shape` neighborhood to `NaN` (avoiding pixels that are outside the ellipsoid). Not a trivial operation. If we don't want `NaN` support, I would completely disallow images containing `NaN`.

`nansafe` is perfectly fine with me, removing it is, too. How should we proceed? What do the other members think?

Reply:
> You can't simply replace them and bring them back to propagate. You'd have to set all affected pixels in the kernel.shape neighborhood to NaN (avoiding pixels that are outside the ellipsoid). Not a trivial operation. If we don't want NaN support, I would completely disallow images containing NaN.

No, that's not what I was suggesting, I was suggesting (wlog wrt inverse):
- replace the nans with iinfo.maxval, keeping a record of where the nans were
- apply the algorithm
- in the result, put back the nans at the same positions

Actually, since it's the background, it doesn't much matter — could just leave it at 0?

@scikit-image/core especially @stefanv and @hmaarrfk:

- (1) `nansafe=False`, or
- (2) no support for nans, but with a clear guide about how to deal with nans?

I have a slight preference for (1).

Reply:
You would place the `NaN`s into the resulting image or into the background?

Let's take this simple example:

```
img = np.full((5,5), 5, dtype=np.uint8)
img[0, 1] = np.nan
rolling_ellipsoid(img, kernel_shape=(3,3), intensity_vertex=0.1)
```
What would the expected output be?

Reply:
For me at least, I would expect the same result as without the nan, but with a nan in that position. Ie nans are masked.

Reply:
@jni

```
[
[0, np.nan, 0, 0, 0],
[0, 0, 0, 0, 0],
[0, 0, 0, 0, 0],
[0, 0, 0, 0, 0],
[0, 0, 0, 0, 0],
]
```

?